File: 2026-01/src/day_02/boj_10866/Main.java
package day_02.boj_10866;

import java.io.*;
import java.util.*;

public class Main {

    static final int MAX = 20000;
    static int[] deque = new int[MAX];
    static int front = MAX / 2, rear = MAX / 2;

    static void push_front(int x) {
        deque[--front] = x;
    }

    static void push_back(int x) {
        deque[rear++] = x;
    }

    static int pop_front() {
        return deque[front++];
    }

    static int pop_back() {
        return deque[--rear];
    }

    static int size() {
        return rear - front;
    }

    static boolean empty() {
        return front == rear;
    }

    static int front() {
        return deque[front];
    }

    static int back() {
        return deque[rear - 1];
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st;

        int n = Integer.parseInt(br.readLine());
        while (n-- > 0) {
            st = new StringTokenizer(br.readLine());
            String q = st.nextToken();

            if (q.equals("push_front")) {
                int x = Integer.parseInt(st.nextToken());
                push_front(x);
            } else if (q.equals("push_back")) {
                int x = Integer.parseInt(st.nextToken());
                push_back(x);
            } else if (q.equals("pop_front")) {
                if (empty()) {
                    sb.append("-1\n");
                } else {
                    sb.append(pop_front()).append("\n");
                }
            } else if (q.equals("pop_back")) {
                if (empty()) {
                    sb.append("-1\n");
                } else {
                    sb.append(pop_back()).append("\n");
                }
            } else if (q.equals("size")) {
                sb.append(size()).append("\n");
            } else if (q.equals("empty")) {
                if (empty()) {
                    sb.append("1\n");
                } else {
                    sb.append("0\n");
                }
            } else if (q.equals("front")) {
                if (empty()) {
                    sb.append("-1\n");
                } else {
                    sb.append(front()).append("\n");
                }
            } else {
                if (empty()) {
                    sb.append("-1\n");
                } else {
                    sb.append(back()).append("\n");
                }
            }
        }

        System.out.println(sb);
    }
}

File: 2026-01/src/day_02/boj_10866/Main.cpp
#include <bits/stdc++.h>
using namespace std;

const int MAX = 20000;
int dq[MAX];
int s = MAX / 2, e = MAX / 2;

void push_front(int x) {
    dq[--s] = x;
}

void push_back(int x) {
    dq[e++] = x;
}

void pop_front() {
    s++;
}

void pop_back() {
    e--;
}

int size() {
    return e - s;
}

bool empty() {
    return s == e;
}

int front() {
    return dq[s];
}

int back() {
    return dq[e - 1];
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    int n;
    cin >> n;

    while (n--) {
        string q;
        cin >> q;

        if (q == "push_front") {
            int x;
            cin >> x;
            push_front(x);
        } else if (q == "push_back") {
            int x;
            cin >> x;
            push_back(x);
        } else if (q == "pop_front") {
            if (empty()) {
                cout << -1 << '\n';
            } else {
                cout << front() << '\n';
                pop_front();
            }
        } else if (q == "pop_back") {
            if (empty()) {
                cout << -1 << '\n';
            } else {
                cout << back() << '\n';
                pop_back();
            }
        } else if (q == "size") {
            cout << size() << '\n';
        } else if (q == "empty") {
            cout << empty() << '\n';
        } else if (q == "front") {
            if (empty()) {
                cout << -1 << '\n';
            } else {
                cout << front() << '\n';
            }
        } else {
            if (empty()) {
                cout << -1 << '\n';
            } else {
                cout << back() << '\n';
            }
        }
    }
}
